### PR TITLE
linter: warn if expect_num_outputs is missing

### PR DESCRIPTION
**Pull request by bernt-matthias**
if outputs have filters or are discovered

## How to test the changes?
(Select all options that apply)
- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these contributions under [Galaxy's current license](https://github.com/galaxyproject/galaxy/blob/dev/LICENSE.txt).
- [x] I agree to allow the Galaxy committers to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT). If this condition is an issue, uncheck and just let us know why with an e-mail to galaxy-committers@lists.galaxyproject.org.

### CODE REVIEW

**Review by mvdbeek — COMMENTED** [review]
Can you push this into a profile version ?
I don't think this should be a warning for old tools.

**Comment by bernt-matthias** [comment]
> Can you push this into a profile version ? I don't think this should be a warning for old tools.

Hrmm. This change should not affect the functionality of tools, but just the linter. Which is as far as I see only called from `planemo`... imho the linter should always output according to the latest standards which allows tools to slowly catch up (always if they are touched).

File: lib/galaxy/tool_util/linters/tests.py [hunk 1]
```diff
@@ -26,6 +26,12 @@ def lint_tsts(tool_xml, lint_ctx):
                 has_test = True
                 break
 
+        # check if expect_num_outputs is set if there are outputs with filters or discovered datasets
+        discover_datasets = tool_xml.findall("./outputs//discover_datasets")
+        filter = tool_xml.findall("./outputs//filter")
+        if len(discover_datasets) + len(filter) > 0 and "expect_num_outputs" not in test.attrib:
+            lint_ctx.warn("Test should specify 'expect_num_outputs' if outputs have filters or use discover_datasets")
```
Comment (mvdbeek) [inline lib/galaxy/tool_util/linters/tests.py]: There was a discussion about this, maybe we should first do some consistent counting of collections and datasets and describe that.

Comment (bernt-matthias) [reply lib/galaxy/tool_util/linters/tests.py]: Adapted it to the latest state of https://github.com/galaxyproject/galaxy/pull/7894

**Comment by nsoranzo** [comment]
@bernt-matthias If ready for review, please remove the `status/WIP` label.

File: lib/galaxy/tool_util/linters/tests.py [hunk 2]
```diff
@@ -38,11 +38,16 @@ def lint_tsts(tool_xml, lint_ctx):
             if len(assertions) == 0:
                 continue
             if len(assertions) > 1:
-                lint_ctx.error(f"Test {test_idx}: More than one {ta} found. Only the first is considered.")
+                lint_ctx.error(f"Test {test_idx}: More than one {ta} found. Only the first is considered.", node=test)
```
Comment (jmchilton) [inline lib/galaxy/tool_util/linters/tests.py:41]: This should be a warn, but unrelated change.

Comment (gallardoalba) [reply lib/galaxy/tool_util/linters/tests.py:41]: I think it does not allow to include of more than one assertion per test, which I'm not sure if it is really practical.

Comment (bernt-matthias) [reply lib/galaxy/tool_util/linters/tests.py:41]:
```suggestion
                lint_ctx.warn(f"Test {test_idx}: More than one {ta} found. Only the first is considered.", node=test)
```

Comment (bernt-matthias) [reply lib/galaxy/tool_util/linters/tests.py:41]: Thanks for the suggestion @jmchilton .. should I apply it in a separate PR?

@gallardoalba .. its that you can have only one `assert_command` (`assert_stderr`, or `assert_stdout`) per test. Within these you can have an arbitrary number of assertions.

Comment (gallardoalba) [reply lib/galaxy/tool_util/linters/tests.py:41]: Ahhh, I understand. Thanks a lot!
